File: 1]. DSA + CP/1]. DSA/2]. Algorithms/01]. Searching Algorithms/Python/01) Binary Search.py
def binary_search(array, target):
    first = 0
    last = len(array) - 1

    while first <= last:
        midpoint = (first + last) // 2

        if array[midpoint] == target:
            return midpoint
        elif array[midpoint] < target:
            first = midpoint + 1
        elif array[midpoint] > target:
            last = midpoint - 1

    return None


def verify(index, target):
    if index is not None:
        print("Target", target, "found at index:", index)
    else:
        print("Target", target, "not in list")


array = [x for x in range(1, 11)]
print("Input array:", array)

verify(binary_search(array, 6), 6)
verify(binary_search(array, 20), 20)

File: 1]. DSA + CP/1]. DSA/2]. Algorithms/01]. Searching Algorithms/C++/01) Binary Search.cpp
#include <iostream>
using namespace std;
int binarySearch (int arr[], int n, int key)
{
    int s = 0, e = n - 1;
    while (s <= e)
    {
        int mid = s + (e - s) / 2;
        if (arr[mid] == key)
	{
	    return mid;
	}
        if (key > arr[mid])
	{
	    s = mid + 1;
	}
        else
	{
	    e = mid - 1;
	}
        mid = s + (e - s) / 2;
    }
    return -1;
}

int main ()
{
    int arr[] = { 1, 2, 4, 5, 7, 9, 11 };
    int n = 7;
    int ans = binarySearch (arr, 7, 2);
    if (ans == -1)
        cout << "Not Found ";
    else
        cout << " Element found at index " << ans;
}
